pkg/cache/unavailableofferings_test.go

```diff
@@ -18,43 +18,193 @@ package cache
+		Name:         &name,
+		Family:       &family,
+		Size:         &name,
+		Capabilities: &[]compute.ResourceSkuCapabilities{{Name: lo.ToPtr(skewer.VCPUs), Value: lo.ToPtr(strconv.Itoa(cpuCount))}},
+	}
+}
+
+func assertOfferingUnavailable(t *testing.T, u *UnavailableOfferings, sku *skewer.SKU, zone, capacityType, message string) {
+	t.Helper()
+	if !u.IsUnavailable(sku, zone, capacityType) {
+		t.Errorf("%s: %s", sku.GetName(), message)
+	}
+}
+
+func assertOfferingAvailable(t *testing.T, u *UnavailableOfferings, sku *skewer.SKU, zone, capacityType, message string) {
+	t.Helper()
+	if u.IsUnavailable(sku, zone, capacityType) {
+		t.Errorf("%s: %s", sku.GetName(), message)
+	}
+}
+
 func TestUnavailableOfferings(t *testing.T) {
 	// create a new cache with a short TTL
-	c := cache.New(time.Second, time.Second)
-	u := NewUnavailableOfferingsWithCache(c)
+	singleInstanceCache := cache.New(time.Second, time.Second)
+	vmFamilyCache := cache.New(time.Second, time.Second)
+	u := NewUnavailableOfferingsWithCache(singleInstanceCache, vmFamilyCache)
+	testSKU := createTestSKU("NV16as_v4", "NVasv4Family", 16)
 
 	// test that an offering is not marked as unavailable initially
-	if u.IsUnavailable("NV16as_v4", "westus", "spot") {
-		t.Error("Offering should not be marked as unavailable initially")
-	}
+	assertOfferingAvailable(t, u, testSKU, "westus", karpv1.CapacityTypeSpot, "Offering should not be marked as unavailable initially")
 
 	// mark the offering as unavailable
-	u.MarkUnavailableWithTTL(context.TODO(), "test reason", "NV16as_v4", "westus", "spot", time.Second)
+	u.MarkUnavailableWithTTL(context.TODO(), "test reason", "NV16as_v4", "westus", karpv1.CapacityTypeSpot, time.Second)
 
 	// test that the offering is now marked as unavailable
-	if !u.IsUnavailable("NV16as_v4", "westus", "spot") {
-		t.Error("Offering should be marked as unavailable after being marked as such")
-	}
+	assertOfferingUnavailable(t, u, testSKU, "westus", karpv1.CapacityTypeSpot, "Offering should be marked as unavailable after being marked as such")
+	// test that offering is available for different capacity type
+	assertOfferingAvailable(t, u, testSKU, "westus", karpv1.CapacityTypeOnDemand, "Offering should be available for different capacity type")
 
 	// wait for the cache entry to expire
 	time.Sleep(time.Second)
 
 	// test that the offering is no longer marked as unavailable
-	if u.IsUnavailable("NV16as_v4", "westus", "spot") {
-		t.Error("Offering should not be marked as unavailable after cache entry has expired")
+	assertOfferingAvailable(t, u, testSKU, "westus", karpv1.CapacityTypeSpot, "Offering should not be marked as unavailable after cache entry has expired")
+}
+
+func getVersionedSKUFamily(sku *skewer.SKU) string {
+	skuVMSize, _ := sku.GetVMSize()
+	skuVersion := "1"
+	if skuVMSize.Version != "" {
+		skuVersion = skuVMSize.Version[1:]
+	}
+	return skuVMSize.Family + skuVersion
+}
+
+func TestUnavailableOfferingsVMFamilyLevel(t *testing.T) {
```

Comment:
minor: Consider naming this test a bit better?

IMO tests should say what they are doing and what they expect to happen (we don't follow that consistently in the Karpenter codebase).

I think this is actually two tests:
`TestUnavailableOfferings_VMFamilyCoreLimit_AllowsFewerCores`
and
`TestUnavailableOfferings_VMFamily_BlocksAll`

(or something -- free free to fiddle with the names but I think you get the idea).

Reply:
Yeah good idea to split these two, initially I just included all of the "VM family cache" functionality in that one test. I separated them into two distinct ones, kept the names you suggested since they seemed to explain everything properly.